transports/webrtc-websys/src/upgrade.rs

```diff
@@ -0,0 +1,124 @@
+        .expect("certificate to be valid");
+
+    let certificate = JsFuture::from(certificate_promise).await?; // Needs to be Send
+
+    let mut config = RtcConfiguration::default();
+    // wrap certificate in a js Array first before adding it to the config object
+    let certificate_arr = js_sys::Array::new();
+    certificate_arr.push(&certificate);
+    config.certificates(&certificate_arr);
+
+    let peer_connection = web_sys::RtcPeerConnection::new_with_configuration(&config)?;
+
+    // Create substream for Noise handshake
+    // Must create data channel before Offer is created for it to be included in the SDP
+    let handshake_data_channel = RtcDataChannelBuilder::default()
+        .negotiated(true)
+        .build_with(&peer_connection);
+
+    let webrtc_stream = Stream::new(handshake_data_channel);
+
+    let ufrag = format!("libp2p+webrtc+v1/{}", gen_ufrag(32));
+    /*
+     * OFFER
+     */
+    let offer = JsFuture::from(peer_connection.create_offer()).await?; // Needs to be Send
+    let offer_obj = sdp::wasm::offer(offer, &ufrag);
+    log::trace!("Offer SDP: {:?}", offer_obj);
+    let sld_promise = peer_connection.set_local_description(&offer_obj);
+    JsFuture::from(sld_promise)
+        .await
+        .expect("set_local_description to succeed");
+
+    /*
+     * ANSWER
+     */
+    // TODO: Update SDP Answer format for Browser WebRTC
+    let answer_obj = sdp::wasm::answer(sock_addr, &remote_fingerprint, &ufrag);
+    log::trace!("Answer SDP: {:?}", answer_obj);
+    let srd_promise = peer_connection.set_remote_description(&answer_obj);
+    JsFuture::from(srd_promise)
+        .await
+        .expect("set_remote_description to succeed");
+
+    // get local_fingerprint from local RtcPeerConnection peer_connection certificate
+    let local_sdp = match &peer_connection.local_description() {
+        Some(description) => description.sdp(),
+        None => return Err(Error::JsError("local_description is None".to_string())),
+    };
+    let local_fingerprint = match sdp::fingerprint(&local_sdp) {
+        Some(fingerprint) => fingerprint,
+        None => return Err(Error::JsError(format!("No local fingerprint found"))),
+    };
+
+    log::trace!("local_fingerprint: {:?}", local_fingerprint);
+    log::trace!("remote_fingerprint: {:?}", remote_fingerprint);
+
+    let peer_id = noise::outbound(
+        id_keys,
+        webrtc_stream,
+        remote_fingerprint,
+        local_fingerprint,
+    )
+    .await?;
+
+    log::debug!("Remote peer identified as {peer_id}");
+
+    Ok((peer_id, Connection::new(peer_connection)))
+}
+
+fn gen_ufrag(len: usize) -> String {
+    let charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+    let mut ufrag = String::new();
+    let mut buf = vec![0; len];
+    getrandom::getrandom(&mut buf).unwrap();
+    for i in buf {
+        let idx = i as usize % charset.len();
+        ufrag.push(charset.chars().nth(idx).unwrap());
+    }
+    ufrag
+}
```

Comment:
This is missing the prefix, right?

``` rust

/// Generates a random ufrag and adds a prefix according to the spec.
fn random_ufrag() -> String {
    format!(
        "libp2p+webrtc+v1/{}",
        thread_rng()
            .sample_iter(&Alphanumeric)
            .take(64)
            .map(char::from)
            .collect::<String>()
    )
}
```

Also can this be moved to `webrtc-utils` and used by both native and wasm?

Reply:
Yes you are right! Fixed and Moved.